integration_test.go

```diff
@@ -67,36 +60,54 @@ func runsOnEmulator() bool {
 	return false
 }
 
-func initEmulator() error {
+func initTestInstance(config string) (cleanup func(), err error) {
 	ctx := context.Background()
 	instanceAdmin, err := instance.NewInstanceAdminClient(ctx)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer instanceAdmin.Close()
+	// Check if the instance exists or not.
+	_, err = instanceAdmin.GetInstance(ctx, &instancepb.GetInstanceRequest{
+		Name: fmt.Sprintf("projects/%s/instances/%s", projectId, instanceId),
+	})
+	if err == nil {
+		return func() {}, nil
+	}
+	if spanner.ErrCode(err) != codes.NotFound {
+		return nil, err
+	}
+
+	// Instance does not exist. Create a temporary instance for this test run.
+	// The instance will be deleted after the test run.
 	op, err := instanceAdmin.CreateInstance(ctx, &instancepb.CreateInstanceRequest{
 		Parent:     fmt.Sprintf("projects/%s", projectId),
 		InstanceId: instanceId,
 		Instance: &instancepb.Instance{
-			Config:      fmt.Sprintf("projects/%s/instanceConfigs/%s", projectId, "emulator-config"),
+			Config:      fmt.Sprintf("projects/%s/instanceConfigs/%s", projectId, config),
 			DisplayName: instanceId,
 			NodeCount:   1,
 		},
 	})
 	if err != nil {
-		// Check if the instance has already been created by another (parallel) test.
-		code := spanner.ErrCode(err)
-		if code != codes.AlreadyExists {
-			return fmt.Errorf("could not create instance %s: %v", fmt.Sprintf("projects/%s/instances/%s", projectId, instanceId), err)
-		}
+		return nil, fmt.Errorf("could not create instance %s: %v", fmt.Sprintf("projects/%s/instances/%s", projectId, instanceId), err)
 	} else {
 		// Wait for the instance creation to finish.
 		_, err := op.Wait(ctx)
 		if err != nil {
-			return fmt.Errorf("waiting for instance creation to finish failed: %v", err)
+			return nil, fmt.Errorf("waiting for instance creation to finish failed: %v", err)
 		}
 	}
-	return nil
+	// Delete the instance after all tests have finished.
+	return func() {
+		instanceAdmin, err := instance.NewInstanceAdminClient(ctx)
+		if err != nil {
+			return
```

Comment:
Would this not result in the instance being leaked? If so, should we add code somewhere that cleans up old instances similar to Python and Node.js?

Reply:
I think that the chance that this error will return an error is extremely small, but a more realistic scenario might be that the test run itself times out / panics or in other ways fails and never reaches this point. So your point about cleaning up stale resources is a good one. I've added logic for that as well.